yesod-core/src/Yesod/Core/Types.hs

```diff
@@ -303,6 +305,20 @@ newtype RepXml = RepXml Content
 
 type ContentType = ByteString -- FIXME Text?
 
+-- | Wrapper around types so that Handlers can return a domain type, even when
+-- the data will eventually be encoded as JSON.
+-- Example usage in a type signature:
+--
+-- > postSignupR :: Handler (JSONResponse CreateUserResponse)
+--
+-- And in the implementation:
+--
+-- > return $ JSONResponse $ CreateUserResponse userId
+--
+-- @since 1.6.14
+data JSONResponse a where
```

Comment:
It's an interesting design, but I would have expected a simpler `newtype` and then putting the `ToJSON` constraints on the instances instead. That removes any runtime overhead (however negligible) from an extra data constructor, but more importantly sticks to simpler constructs. Is there a strong motivation for the GADT approach?

Reply:
Hi Michael - adding the constraints onto the instance definitions is possible, but @MaxGabriel mentioned in #1494 that the errors using the newtype approach are a bit cryptic, and to answer your question from that PR, GADTs does make the errors simpler. 

At the end of the day though, the error between the newtype approach and GADTs approach are both completely understandable. I'm happy to implement it with newtype; let me know which way you'd like to proceed, as I do think it is a nice feature to add.

Reply:
Better error messages sounds like a great reason. I rescind my comment, let's go this way. Thanks for clarifying!